fastapi_websocket_rpc/websocket_rpc_client.py

```diff
@@ -15,6 +15,149 @@
+        if self._websocket is not None:
+            # Case opened, we have something to close.
+            self._websocket.close(code)
+
+    async def handle_exception(self, exception: Exception):
+        try:
+            raise exception
+        # See https://websocket-client.readthedocs.io/en/latest/exceptions.html
+        except websocket._exceptions.WebSocketAddressException:
+            logger.info("websocket address info cannot be found")
+            raise
+        except websocket._exceptions.WebSocketBadStatusException:
+            logger.info("bad handshake status code")
+            raise
+        except websocket._exceptions.WebSocketConnectionClosedException:
+            logger.info("remote host closed the connection or some network error happened")
+            raise
+        except websocket._exceptions.WebSocketPayloadException:
+            logger.info(
+                f"WebSocket payload is invalid")
+            raise
+        except websocket._exceptions.WebSocketProtocolException:
+            logger.info(f"WebSocket protocol is invalid")
+            raise
+        except websocket._exceptions.WebSocketProxyException:
+            logger.info(f"proxy error occurred")
+            raise
+        except OSError as err:
+            logger.info("RPC Connection failed - %s", err)
+            raise
+        except Exception as err:
+            logger.exception("RPC Error")
+            raise
+
+class WebSocketsClientHandler(SimpleWebSocket):
+    """
+    Handler that use https://websockets.readthedocs.io/en/stable module.
+    This implementation does not support HTTP proxy (see https://github.com/python-websockets/websockets/issues/364).
+    """
+    def __init__(self):
+        self._websocket = None
+
+    """
+    Args:
+        **kwargs: Additional args passed to connect
+            https://websockets.readthedocs.io/en/stable/reference/asyncio/client.html#opening-a-connection
+    """
+    async def connect(self, uri: str, **connect_kwargs):
+        self._websocket = await websockets.connect(uri, **connect_kwargs)
+
+    async def send(self, msg):
+        if self._websocket is None:
+            # connect must be called before.
+            logging.error("Websocket connect() must be called before.")
+        await self._websocket.send(msg)
+
+    async def recv(self):
+        if self._websocket is None:
+            # connect must be called before.
+            logging.error("Websocket connect() must be called before.")
+        try:
+            msg = await self._websocket.recv()
+        except websockets.exceptions.ConnectionClosed as err:
+            raise BaseConnectionClosedException from err
+        return msg
+
+    async def close(self, code: int = 1000):
+        if self._websocket is not None:
+            # Case opened, we have something to close.
+            self._websocket.close(code)
+
+    async def handle_exception(self, exception: Exception):
+        try:
+            raise exception
```

Comment:
This is actually worse performance-wise - catching and then re raising the exception.
What I meant was to have the __connect__ method be polymorphic (or just a portion of it)

Reply:
So instead of having the `handle_exception` method which starts with `raise exception`.

You have the polymorphism in the `connect` method itself -


```python

class WebSocketsClientHandler(SimpleWebSocket):
    #.... rest of code
    
    # Connect with type specific exceptions
    async def connect(self, uri: str, **connect_kwargs):
        try:
            self._websocket = await websockets.connect(uri, **connect_kwargs)
        except ConnectionRefusedError:
            logger.info("RPC connection was refused by server")
            raise
        except ConnectionClosedError:
            logger.info("RPC connection lost")
            raise
        except ConnectionClosedOK:
            logger.info("RPC connection closed")
            raise
        except InvalidStatusCode as err:
            logger.info(
                f"RPC Websocket failed - with invalid status code {err.status_code}")
            raise
        except WebSocketException as err:
            logger.info(f"RPC Websocket failed - with {err}")
            raise  

class WebSocketClientHandler(SimpleWebSocket):
    #.... rest of code
    
    # Connect with type specific exceptions
    async def connect(self, uri: str, **connect_kwargs):
        try:
           self._websocket = await asyncio.get_event_loop().run_in_executor(None, websocket.create_connection, uri, **connect_kwargs)
        # See https://websocket-client.readthedocs.io/en/latest/exceptions.html
        except websocket._exceptions.WebSocketAddressException:
            logger.info("websocket address info cannot be found")
            raise
        except websocket._exceptions.WebSocketBadStatusException:
            logger.info("bad handshake status code")
            raise
        except websocket._exceptions.WebSocketConnectionClosedException:
            logger.info("remote host closed the connection or some network error happened")
            raise
        except websocket._exceptions.WebSocketPayloadException:
            logger.info(
                f"WebSocket payload is invalid")
            raise
        except websocket._exceptions.WebSocketProtocolException:
            logger.info(f"WebSocket protocol is invalid")
            raise
        except websocket._exceptions.WebSocketProxyException:
            logger.info(f"proxy error occurred")
            raise


# With the generic excpetions (OSError , Exception) being handled in the wrapping  Class - WebSocketRpcClient

async def __connect__(self):
        try:
           # .... actual code

        except OSError as err:
            logger.info("RPC Connection failed - %s", err)
            raise
        except Exception as err:
            logger.exception("RPC Error")
            raise

```

Reply:
Oh I see, you meant to put the exception handling inside the connect function. Yes no problem with that! Will do.

Reply:
Done!

Reply:
https://github.com/antoinetran/fastapi_websocket_rpc/commit/54e12efd4353ffa1a0f94bbc97e041385f9771c4